src/machine/machine_stm32_spi.go

```diff
@@ -17,83 +39,178 @@ type SPIConfig struct {
 	SDI       Pin
 	LSBFirst  bool
 	Mode      uint8
+	Direction uint8
+	DataSize  uint8
+	CSMode    uint8
 }
 
 // Configure is intended to setup the STM32 SPI1 interface.
-// Features still TODO:
-// - support SPI2 and SPI3
-// - allow setting data size to 16 bits?
-// - allow setting direction in HW for additional optimization?
-// - hardware SS pin?
 func (spi SPI) Configure(config SPIConfig) {
+
+	// -- CONFIGURING THE SPI IN MASTER MODE --
+	//
+	// 1. Select the BR[2:0] bits to define the serial clock baud rate (see
+	// 	  SPI_CR1 register).
+	// 2. Select the CPOL and CPHA bits to define one of the four relationships
+	//    between the data transfer and the serial clock (see Figure 248). This
+	//    step is not required when the TI mode is selected.
+	// 3. Set the DFF bit to define 8- or 16-bit data frame format
+	// 4. Configure the LSBFIRST bit in the SPI_CR1 register to define the frame
+	//    format. This step is not required when the TI mode is selected.
+	// 5. If the NSS pin is required in input mode, in hardware mode, connect the
+	//    NSS pin to a high-level signal during the complete byte transmit
+	//    sequence. In NSS software mode, set the SSM and SSI bits in the SPI_CR1
+	//    register. If the NSS pin is required in output mode, the SSOE bit only
+	//    should be set. This step is not required when the TI mode is selected.
+	// 6. Set the FRF bit in SPI_CR2 to select the TI protocol for serial
+	//    communications.
+	// 7. The MSTR and SPE bits must be set (they remain set only if the NSS pin
+	//    is connected to a high-level signal).
+
+	// disable SPI interface before any configuration changes
+	spi.Bus.CR1.ClearBits(stm32.SPI_CR1_SPE)
+
 	// enable clock for SPI
 	enableAltFuncClock(unsafe.Pointer(spi.Bus))
 
+	// init pins
+	if config.SCK == 0 && config.SDO == 0 && config.SDI == 0 {
+		config.SCK = SPI0_SCK_PIN
+		config.SDO = SPI0_SDO_PIN
+		config.SDI = SPI0_SDI_PIN
+	}
+	spi.configurePins(config)
+
 	// Get SPI baud rate based on the bus speed it's attached to
 	var conf uint32 = spi.getBaudRate(config)
 
 	// set bit transfer order
 	if config.LSBFirst {
 		conf |= stm32.SPI_CR1_LSBFIRST
+	} else {
+		conf &^= stm32.SPI_CR1_LSBFIRST
```

Comment:
Why do you clear the bits here and below? `getBaudRate` should have all those bits zeroed.

A few days ago I used a very similar pattern here (and in other places in the PR):
https://github.com/tinygo-org/tinygo/pull/1379/files#diff-cb056e43ce0e9bc6e837cd32b2b9ac66R352-R357

Reply:
Right, the clearing isn't necessary. At one point I had separate functions for these individual bits in `CR1` (so the user could enable/disable features individually at runtime). Tried to simplify the interface and merged them altogether back into this sole `Configure` function. Will remove the clearing statements. (Are they not optimized away?)